spec/compromise-syntax.md

```diff
@@ -0,0 +1,238 @@
+```
+{$name}
+{$count :number}
+{$fraction :number style=percent minFractions=2}
+{<25> :number}
+{:specialFunction optionKey=optionValue key2=<value with spaces>}
+```
+
+An argument name is a `$` immediately followed by an identifier.
+A message formatting function will typically accept a Map of argument keys to values
+where the keys match argument name identifiers in the patterns of the message.
+
+TODO: For the definition of identifiers we should consult with the Unicode Source Code Working Group.
+
+If the placeholder specifies only an argument name,
+then the formatting function is inferred from the run-time type of the argument value.
+For example, a string value would simply be inserted,
+and a numeric type could be formatted using some kind of default number formatter.
+- TODO: In the registry, specify the default formatters for a small set of value types.
+
+The function is specified via a `:` immediately followed by an identifier.
+If an argument name or a value literal is given,
+then the function is usually a formatter for its expected input types.
+- TODO: There still seems to be discussion about the function prefix character.
+  It could be some other ASCII punctuation, for example `@`.
+- TODO: Functions must be listed in a registry.
+- TODO: Functions that accept value literals must specify their syntax.
+- TODO: Reserve a naming convention for private use functions (not in the standard registry). Examples:
+  - Starts with `_`
+  - Starts with `x`
+  - Contains interior dots – e.g., com.google.fancyNumber
+
+When a function is specified, it can be optionally followed by options which are key-value pairs,
+with `=` (and no white space) between the key identifier and the value.
+The option value can contain any character other than curly braces and white space,
+unless delimited like literal values.
+- TODO: Each registered function must define the available options and their value syntax.
+- TODO: If we allow white space in option values, then we need optional delimiters for such values. Probably the same delimiters as for literal values.
+
+Options are not allowed when no function is specified.
+
+Value literals are important for developers to control the output.
+For example, certain strings may need to be inlined as literals so that
+they are not changed during translation.
+Numeric constants need to be formatted differently depending on the target language
+(e.g., which digits and separators, and the grouping style).
+Date constants need to be formatted according to the target language’s calendar system.
+
+If only a value literal is given, without specifying a function,
+then its string value is used verbatim and it is read-only for translators.
+- TODO: Value literals need to be delimited (they may contain spaces),
+  and the starting delimiter needs to be distinct from the prefixes for
+  argument names and functions.
+  Reasonable choices include `<>`, `()`, `[]`, or `||`.
```

Comment:
```
I would also include U+0060	GRAVE ACCENT as a possibility, eg {`25` :number}
```
It is a very infrequent character in common text messages.

Reply:
Ok, I added the grave accent. (I am not sure how to use Markdown to show a pair of grave accents in code style...)

However, note that for a *list* of space-separated value literals, as may occur in a selector variant, it would be visually better to use open-close delimiters. That probably suggests using `()`.